### ISSUE TITLE
Quest: Improve Code quality

### ISSUE THREAD

**Issue by stefanpenner** [ember-cli/ember-cli#3730]
I've enabled code climate, it's likely a good idea that we attempt to improve some of the problematic areas. Some things, appear invalid (browser checks, in node code etc), we should report them with code-climate and help them improve.

We are in a relatively good place was(2.2) -> now [![Code Climate](https://codeclimate.com/github/ember-cli/ember-cli/badges/gpa.svg)](https://codeclimate.com/github/ember-cli/ember-cli) but we can strive to improve further.

Anyways, hopefully we can collectively triage some of the problem areas.

- [ ] D	[lib/broccoli/ember-app.js](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/ember-app.js)
- [ ] D	[lib/models/blueprint.js](https://github.com/ember-cli/ember-cli/blob/master/lib/models/blueprint.js)
- [ ] D	[lib/tasks/npm-install.js](https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/npm-install.js)
- [ ] D	[lib/tasks/npm-uninstall.js](https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/npm-uninstall.js)
- [ ] D	[lib/models/project.js](https://github.com/ember-cli/ember-cli/blob/master/lib/models/project.js)
- [ ] D	[lib/models/addon.js](https://github.com/ember-cli/ember-cli/blob/master/lib/models/addon.js)
- [ ] C	[lib/models/command.js](https://github.com/ember-cli/ember-cli/blob/master/lib/models/command.js)
- [ ] B	[blueprints/component-addon/index.js](https://github.com/ember-cli/ember-cli/blob/master/blueprints/component-addon/index.js)
- [ ] B	[blueprints/component/index.js](https://github.com/ember-cli/ember-cli/blob/master/blueprints/component/index.js)
- [ ] B	[lib/commands/install-bower.js](https://github.com/ember-cli/ember-cli/blob/master/lib/commands/install-bower.js)
- [ ] B	[lib/commands/install-npm.js](https://github.com/ember-cli/ember-cli/blob/master/lib/commands/install-npm.js)
- [ ] B	[lib/commands/uninstall-npm.js](https://github.com/ember-cli/ember-cli/blob/master/lib/commands/uninstall-npm.js)

[edit: updated list – Fri May 22 09:45:01 PDT 2015]

source: https://codeclimate.com/github/ember-cli/ember-cli/code?sort=remediation_cost&sort_direction=desc

Pending tasks:

- [x] figure out why PR notification isn't working (i followed the steps..)

  this is a feature not currently available for public OSS Projects, @codeclimate is currently working on this feature, and I will push to have us on the beta if at all possible :)

future questions:

- [tests are currently not being rated, as they do incure a duplication penalty. I do not personally feel that the duplication in our tests is poor, rather I believe it to be the right level of abstraction. How do we proceed?
-  how do we handle new syntax, that codeclimate does not yet support? (likely something to bring up with them)

future tasks:

- [ ] enable code-coverage:
  - [x] for ember-cli -> #3795
  - [ ] for add-ons by default
  - [ ] for apps by default

**Comment by kellyselden** [comment]
Great idea. I will help look into this.

**Comment by stefanpenner** [comment]
@kellyselden got any more time ;)

**Comment by kellyselden** [comment]
@stefanpenner Not for a few weeks. But I will keep it on my radar.

**Comment by joostdevries** [comment]
did no see this but :heart_eyes: I'll try to help out.

**Comment by chrislopresto** [comment]
@stefanpenner I just forked and ran Code Climate. I'm getting a 2.1 score rather than the 3.1 master has.

I verified the commits match, and a spot check shows files with the exact same analysis you have. Is there a chance you entered project settings with file/directory exclusions via the web UI? If so, we can now persist them in a `.codeclimate.yml` file at the repository root. http://docs.codeclimate.com/article/289-configuring-your-repository-via-codeclimate-yml
